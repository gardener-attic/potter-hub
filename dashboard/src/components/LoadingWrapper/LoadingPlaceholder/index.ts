import LoadingPlaceholder from "./LoadingPlaceholder";

export default LoadingPlaceholder;
