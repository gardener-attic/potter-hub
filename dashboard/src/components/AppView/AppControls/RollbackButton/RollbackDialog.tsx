import * as React from "react";
import LoadingWrapper from "../../../LoadingWrapper";

import "./RollbackDialog.css";

interface IRollbackDialogProps {
  loading: boolean;
  currentRevision: number;
  onConfirm: (revision: number) => Promise<any>;
  closeModal: () => Promise<any>;
}

interface IRollbackDialogState {
  targetRevision: number;
}

class RollbackDialog extends React.Component<IRollbackDialogProps, IRollbackDialogState> {
  public state: IRollbackDialogState = {
    targetRevision: this.props.currentRevision - 1,
  };

  public render() {
    const options: number[] = [];
    // Use as options the number of versions without the latest
    for (let i = this.props.currentRevision - 1; i > 0; i--) {
      options.push(i);
    }
    return this.props.loading === true ? (
      <div className="row confirm-dialog-loading-info">
        <div className="col-8 loading-legend">Loading, please wait</div>
        <div className="col-4">
          <LoadingWrapper />
        </div>
      </div>
    ) : (
      <div>
        <label>
          Select the revision to which you want to rollback (current: {this.props.currentRevision})
        </label>
        <div>
          <select className="margin-t-normal" onChange={this.selectRevision}>
            {options.map(o => (
              <option key={o} value={o}>
                {o}
              </option>
            ))}
          </select>
        </div>
        <div className="margin-t-normal button-row">
          <button className="button" type="button" onClick={this.props.closeModal}>
            Cancel
          </button>
          <button
            className="button button-primary button-danger"
            type="submit"
            onClick={this.onConfirm}
          >
            Rollback
          </button>
        </div>
      </div>
    );
  }

  private selectRevision = (e: React.FormEvent<HTMLSelectElement>) => {
    this.setState({ targetRevision: Number(e.currentTarget.value) });
  };

  private onConfirm = () => {
    this.props.onConfirm(this.state.targetRevision);
  };
}

export default RollbackDialog;
