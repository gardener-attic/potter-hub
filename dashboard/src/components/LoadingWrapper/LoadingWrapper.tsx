import * as React from "react";

import LoadingSpinner from "../LoadingSpinner";
import LoadingPlaceholder from "./LoadingPlaceholder";

export enum LoaderType {
  Spinner,
  Placeholder,
}

export interface ILoadingWrapperProps {
  type?: LoaderType;
  loaded?: boolean;
  size?: string;
}

// TODO(miguel) Move these kind of components to stateless compontents once we upgrade ts definitions
// Currently, I am having issues transforming it via const LoadingWrapper: React.SFC<ILoadingWrapperProps>
class LoadingWrapper extends React.Component<ILoadingWrapperProps> {
  public static defaultProps: ILoadingWrapperProps = {
    type: LoaderType.Spinner,
    loaded: false,
  };

  public render() {
    return this.props.loaded ? this.props.children : this.renderLoader();
  }

  private renderLoader() {
    switch (this.props.type) {
      case LoaderType.Spinner:
        return <LoadingSpinner size={this.props.size} />;
      case LoaderType.Placeholder:
        return <LoadingPlaceholder />;
    }
    return;
  }
}

export default LoadingWrapper;
