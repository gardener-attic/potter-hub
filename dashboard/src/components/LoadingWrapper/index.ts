import LoadingWrapper, { ILoadingWrapperProps as loadingWrapperProps, LoaderType } from "./LoadingWrapper";
export { LoaderType };
export type ILoadingWrapperProps = loadingWrapperProps;
export default LoadingWrapper;
