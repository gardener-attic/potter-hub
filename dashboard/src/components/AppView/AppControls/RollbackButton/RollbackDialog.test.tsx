import { shallow } from "enzyme";
import * as React from "react";
import LoadingWrapper from "../../../../components/LoadingWrapper";
import RollbackDialog from "./RollbackDialog";

const defaultProps = {
  loading: false,
  currentRevision: 2,
  onConfirm: jest.fn(),
  closeModal: jest.fn(),
};

it("should render the loading view", () => {
  const wrapper = shallow(<RollbackDialog {...defaultProps} loading={true} />);
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find(LoadingWrapper)).toExist();
});

it("should render the form if it is not loading", () => {
  const wrapper = shallow(<RollbackDialog {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find(LoadingWrapper)).not.toExist();
  expect(wrapper.find("select")).toExist();
});

it("should submit the current revision", () => {
  const onConfirm = jest.fn();
  const wrapper = shallow(<RollbackDialog {...defaultProps} onConfirm={onConfirm} />);
  wrapper.setState({ revision: 1 });
  const submit = wrapper.find(".button-danger");
  expect(submit).toExist();
  submit.simulate("click");
  expect(onConfirm).toBeCalledWith(1);
});
