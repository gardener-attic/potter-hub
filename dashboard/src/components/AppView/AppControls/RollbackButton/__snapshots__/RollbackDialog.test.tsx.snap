// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the form if it is not loading 1`] = `
<div>
  <label>
    Select the revision to which you want to rollback (current: 
    2
    )
  </label>
  <div>
    <select
      className="margin-t-normal"
      onChange={[Function]}
    >
      <option
        key="1"
        value={1}
      >
        1
      </option>
    </select>
  </div>
  <div
    className="margin-t-normal button-row"
  >
    <button
      className="button"
      onClick={[MockFunction]}
      type="button"
    >
      Cancel
    </button>
    <button
      className="button button-primary button-danger"
      onClick={[Function]}
      type="submit"
    >
      Rollback
    </button>
  </div>
</div>
`;

exports[`should render the loading view 1`] = `
<div
  className="row confirm-dialog-loading-info"
>
  <div
    className="col-8 loading-legend"
  >
    Loading, please wait
  </div>
  <div
    className="col-4"
  >
    <LoadingWrapper
      loaded={false}
      type={0}
    />
  </div>
</div>
`;
